data/transactions/logic/eval.go

```diff
@@ -4060,3 +4061,62 @@ func opBase64Decode(cx *EvalContext) {
 	}
 	cx.stack[last].Bytes, cx.err = base64Decode(cx.stack[last].Bytes, encoding)
 }
+
+func parseJSON(jsonText []byte) (interface{}, error) {
+	if !json.Valid(jsonText) {
+		return nil, fmt.Errorf("invalid json text %s", jsonText)
+	}
+	var parsed interface{}
+	err := json.Unmarshal(jsonText, &parsed)
+	if err != nil {
+		return nil, err
+	}
+	return parsed, nil
+}
+func opJSONRef(cx *EvalContext) {
+	last := len(cx.stack) - 1
+	fields := strings.Split(string(cx.stack[last].Bytes), ".") // get json keys
+	cx.stack = cx.stack[:last]                                 // pop
+	last = len(cx.stack) - 1
+	resp, err := parseJSON(cx.stack[last].Bytes)
+	if err != nil {
+		cx.err = fmt.Errorf("error while parsing JSON text, %v", err)
+		return
+	}
+	//get value from json
+	for i := 0; i < len(fields); i++ {
+		key := fields[i]
+		switch resp.(type) {
+		case map[string]interface{}:
+			if _, ok := resp.(map[string]interface{})[key]; ok {
+				resp = resp.(map[string]interface{})[key]
+			} else {
+				cx.err = fmt.Errorf("key %s not found in JSON text", key)
+				return
+			}
+		default:
+			cx.err = fmt.Errorf("key %s not found in JSON text", key)
+			return
+		}
+	}
+	expectedType := jsonRefTypeNames[cx.program[cx.pc+1]]
+	var val stackValue
+	switch resp.(type) {
+	case float64:
+		if expectedType != "JSONInt" {
+			cx.err = fmt.Errorf("value retrieved from JSON text is type int but expected %s", expectedType)
+			return
+		}
+		val.Uint = uint64(resp.(float64))
```

Comment:
I don't think we can parse this way.  Very large integers will get converted to a float, and then during the cast to uint64, they will become a different integer. (also, this allows a negative number, which then gets converted to positive by the cast)